src/sgame/sg_bot_skilltree.cpp

```diff
@@ -0,0 +1,208 @@
+
+	// humans
+	{ "predict-aim",        BOT_H_PREDICTIVE_AIM, 5, pred_human, {} },
+};
+
+static const std::vector<botSkillTreeElement_t> initial_unlockable_skills = {
+	// movement skills
+	{ "movement", BOT_B_BASIC_MOVEMENT, 2, pred_always, movement_skills },
+	// fighting skills
+	{ "fighting", BOT_B_BASIC_FIGHT, 2, pred_always, fighting_skills },
+	// situation awareness and survival
+	{ "feels-pain", BOT_B_PAIN, 2, pred_always, survival_skills },
+};
+
+// Note: this function modifies possible_choices to remove the one we just
+//       chose, and add the new unlocked skills to that list.
+static Util::optional<botSkillTreeElement_t> ChooseOneSkill(const gentity_t *bot, skillSet_t skillSet, std::vector<botSkillTreeElement_t> &possible_choices, std::mt19937_64& rng)
+{
+	int total_skill_points = 0;
+	for (const auto &skill : possible_choices)
+	{
+		if (skill.predicate(bot, skillSet))
+		{
+			total_skill_points += skill.cost;
+		}
+	}
+
+	if (total_skill_points <= 0)
+		return Util::nullopt;
+
+	std::uniform_int_distribution<int> dist(0, total_skill_points - 1);
+	int random = dist(rng);
+
+	// find which skill this corresponds to (higher point skills are more
+	// likely to be found)
+	int cursor = 0;
+	for (auto skill = possible_choices.begin(); skill != possible_choices.end(); skill++)
+	{
+		if (!skill->predicate(bot, skillSet))
+		{
+			continue;
+		}
+
+		cursor += skill->cost;
+		if (random < cursor)
+		{
+			botSkillTreeElement_t result = *skill;
+
+			// update the list of future choices
+			possible_choices.erase(skill);
+			for (const auto &child : result.unlocked_skills)
+			{
+				possible_choices.push_back(std::move(child));
+			}
+			result.unlocked_skills.clear();
+
+			return result;
+		}
+	}
+	return Util::nullopt;
+}
+
+// This selects a set of skills for a bot.
+//
+// The algorithm is stable and deterministic.
+// Deterministic:
+//     a bot's skill set is determined only by the bot's name and the bot's
+//     skill level. A different bot will have a different skillset, but this
+//     means that a bot with the same name will have the same skillset from
+//     a game to another.
+// Stable:
+//     when a bot changes skill level, the first skills unlocked are common
+//     between both skill levels. This means that a bot whose skill is raised
+//     by BotSetSkillLevel will keep the skills it already had, and get new
+//     ones.
+//     This essentially means that a bot will keep its "personality" after
+//     a skill change, although keeping its base skills or unlocking new ones.
+std::pair<std::string, skillSet_t> BotDetermineSkills(gentity_t *bot, int skill)
+{
+	std::vector<botSkillTreeElement_t> possible_choices = initial_unlockable_skills;
+	// aliens have 50 points to spend max
+	// humans have 39 points to spend max
+	// here we give a bit more money to humans because they have more
+	// expensive skills
+	float max = G_Team(bot) == TEAM_ALIENS ? 50.0f : 45.0f;
```

Comment:
Do you have any idea on how to maintain these numbers as skill sets change?

Reply:
I'd say there is unfortunately a fair bit of feeling and adjustments. Finding some balance that is about right is also quite important, as we want bots to be about balanced for both teams. Thankfully we don't modify the bot skills that often in theory.

On the flip side, we now have a way of balancing bot strength, which was missing for long.

Reply:
Can we add a static assert or something based on the size of the skills struct or something that will fail to remind us to update this value? I feel like this is the type of thing that people will forget about.

Reply:
Even if the number doesn't end up changing, the static struct will need to be updated, which is the point when new skills are added/removed.

Reply:
Maybe we can add comments everywhere about that, but I'm afraid a static_assert will need some extra function, and possibly be confusing.

One way to do what you suggest is to have a recursive constexpr function that ignores any condition. That's maybe like 10 lines and that can only be useful here. Do you think that's worth it?

Reply:
Hm, I think a comment is sufficient. Probably where you define skills and where the number is defined. In hindsight, a static assert might be pretty annoying.

Reply:
I've added a comment for that in a new commit. I think that should be good.